### DISCUSSION TITLE
Sqlalchemy order_by hybrid_property orders by something else

### DISCUSSION THREAD

**Discussion by kingkong-cmd** [sqlalchemy/sqlalchemy#10770] [Usage Questions]
So, I have a Product class and a User class. A User is the seller of a Product, and can have many products. The Product.price is the price without VAT, and each User have a flag in the User table `is_vat`, to state if the price showed externally shall have VAT added to it or not.

So I wanted to create a query which is ordered by price, and decided to create a hybrid_property (`show_price`) so that I would be able to order_by the price showed externally. The hybrid_property checks if the seller have the `is_vat` flag checked, and then calculates the outward showing price depending on whether VAT should be added or not.

**So, the problem is, the query results come out ordered on `Product.price`, instead of `Product.show_price`, and I don't understand what on earth I am doing wrong?** The query results shows the `show_price` values are calculated correctly.

This is the query I am using (Flask-sqlalchemy & legacy sqlalchemy syntax):

```
products_query = db.session.query(Product)
products = products_query.order_by(Product.show_price.desc()).all()
```

This is the Product class with the hybrid_property:

```
class Product(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    price = db.Column(db.Integer)
    seller_id = db.Column(db.Integer, db.ForeignKey('user.id'))
    
    @hybrid_property
    def show_price(self):
        is_vat = db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0]
        if is_vat == True:
            return (self.price * 1.25)
        if is_vat == False or is_vat == None:
            return (self.price * 1.0)
```
Query results:

```
for i in products:
    print(i.price, i.show_price, i.seller.is_vat)

150 150.0 None
140 175.0 True
140 140.0 None
128 160.0 True
124 155.0 True
120 120.0 False
120 150.0 True
120 150.0 True
120 120.0 None
116 145.0 True
104 130.0 True
100 125.0 True
100 100.0 False
100 125.0 True
95 118.75 True
76 95.0 True
```

**Comment by CaselIT** [comment]
Hi,

The second if path is probably taken that result in `price * 1`.
Also note that what you are doing in the hybrid property is most likely not correct, since you are just taking the value of the fist user returned by `db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0]` and applying it to every product.

**Reply by kingkong-cmd** [reply]
Edit: updated initial post with the query output for more context.

Aha, so the reason I am seeing results for `show_price` is because it is being calculated after the query results have been returned from the DB? Basically it issues N new queries to get `show_price` for each Product object in the query result?

Any idea how i can write the query_property correctly. I initially used `is_vat = self.seller.is_vat` but that gave me an error when I included it in the order_by.

**Reply by CaselIT** [reply]
> Aha, so the reason I am seeing results for show_price is because it is being calculated after the query results have been returned from the DB? Basically it issues N new queries to get show_price for each Product object in the query result?

No, the opposite. when you call `Product.show_price` the query `db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0]` is executed once, the fist user it obtained and the result is used to return the expression.
After the expression is returned the second select is emitted in the external query.

The main issue is that the user retrieved is not connected at all with the product.

> Any idea how i can write the query_property correctly. I initially used is_vat = self.seller.is_vat but that gave me an error when I included it in the order_by.

I'm not sure using a query expression is a good solution here since it leads to a complex query. probably something like this would work
```sql
ORDER BY case (select is_vat from user where id = seller_id limit 1) when true then price * 1.25 else price end
```
(have not tried it)

To avoid such query you could use a function that builds a query with an explicit join

**Reply by kingkong-cmd** [reply]
> No, the opposite. when you call `Product.show_price` the query `db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0]` is executed once, the fist user it obtained and the result is used to return the expression. After the expression is returned the second select is emitted in the external query.
> 
> The main issue is that the user retrieved is not connected at all with the product.
> 

Hmm.. That doesn't quite make sense, but maybe I am missing something of what you are saying. 

The query does in fact produce the `Product.show_price` correctly for each of the products in the query results, ref my output table in my initial post. So at some point the hybrid_property is calculated correctly, however, not sure if it happens simultaneously with the query or if there are N calculations after the query results have been returned from the DB. 

And the `db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0]` is filtered by the seller_id, so it is not the `.first()` of all users (which I think you are implying), it's .first() of one user with `id == seller_id`, i.e. only one user. 

It's the order_by that is not executing as expected, not the query or the `Product.show_price` in itself.

**Reply by CaselIT** [reply]
> And the db.session.query(User.is_vat).filter(User.id == self.seller_id).first()[0] is filtered by the seller_id, so it is not the .first() of all users (which I think you are implying), it's .first() of one user with id == seller_id, i.e. only one user.

That's not how query.first works. Please refer to the documentation https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.first that states that a query is run.

I suggested you also turn on query logging https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.echo

**Reply by kingkong-cmd** [reply]
If the query result only has one row, can't you then use first() to get that row?

**Reply by CaselIT** [reply]
Yes, but fist runs the query, so in this case you get a single value of is_vat for aall products

**Comment by kingkong-cmd** [comment]
I managed to create an expression function to the hybrid property. Apparently, the expression function has to return the value in form of a query object. I have no idea what I am doing, but with the expression function the query results seem to work as expected.

```
@show_price.expression
def show_price(cls):
    return db.session.query(((db.func.coalesce(User.is_vat, False).cast(db.Integer) * 0.25) + 1) * cls.price).where(User.id == cls.seller_id) 
```
